### PR TITLE
feat(orders): BACK-660 Render backorder prompt for reorder/add to shopping list

### PR DESCRIPTION
**Pull request by benpratt77**
Jira: [BACK-660](https://bigcommercecloud.atlassian.net/browse/BACK-660)

## What/Why?
Display backorder prompts for `re-order`/`add to shopping list` modals.

## Rollout/Rollback
Revert this PR if we encounter issues.

## Testing
Tests added/updated.

Note: it is a known/existing thing that a picklist item is merged with an individual purchase of that item and displays as the picklist child's price.
It is also a known/existing thing that the picklist displays as both in the picklist and as an individual item.

Reorder shows fresh inventory stock information:
Screenshot 2026-07-27 at 10 45 50 am

Add to shopping list shows fresh inventory stock information:
Screenshot 2026-07-27 at 10 46 01 am

Parent and child picklist without other children:
Screenshot 2026-07-27 at 10 46 23 am

Still works when no picklists are present:
Screenshot 2026-07-27 at 10 58 42 am

ping @bigcommerce/team-b2b @animesh1987 @declankirk 

[BACK-660]: https://bigcommercecloud.atlassian.net/browse/BACK-660?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

---

> [!NOTE]
> **Medium Risk**
> Adds async inventory and product-search loading to reorder/shopping-list flows and new picklist backorder calculations; mistakes could mislead users on stock or leave actions disabled, but changes are scoped to storefront messaging UI.
> 
> **Overview**
> **Re-order** and **add to shopping list** dialogs now show backorder messaging (ready to ship, backordered qty, restock message) using live catalog inventory, including for **picklist/bundle** parents where child option inventory is shown nested under the parent row.
> 
> `OrderDialog` loads variant info by SKU and, when backorder UI is enabled, resolves picklist selections and calls **`searchProducts`** for those child product IDs. It passes a `picklistProductsById` map into the product list. Dialog inventory loading was tightened so the confirm action is not stuck disabled when the dialog closes, reopens with no products, or the product list clears while a fetch is still in flight.
> 
> `OrderCheckboxProduct` renders per-picklist-option `BackorderMessage` blocks via `getOrderPicklistSelections` and shared catalog backorder helpers. Tests cover picklist nesting, `searchProducts` usage, dialog loading edge cases, and picklist selection resolution.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ce4aa046df5f2406cbc7e642a01dd7a771540f03. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by benpratt77** [comment]
cursor review

File: apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx [hunk 1]
```diff
@@ -508,6 +508,16 @@ export default function OrderDialog({
...
 
+  const catalogInventoryByProductId = useMemo(() => {
+    const map: Record<number, CatalogQuickVariantSku> = {};
+    variantInfoList.forEach((row) => {
+      if (row.productId != null) {
+        map[Number(row.productId)] = row;
+      }
+    });
+    return map;
+  }, [variantInfoList]);
```
[the remark this answers is not in the document]
Comment (benpratt77) [reply apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx]: This was because when the second fetch occurred this caused the display to stutter as the first fetch would complete followed by the second, this was rather unsightly.
Will dig in and see what the validity of this is.
Bearing in mind that there is a merge of products here, the picklist child is merged with the individual purchase if they are the same. I am unsure what happens in the scenario above though.

[the remark this answers is not in the document]
Comment (benpratt77) [reply apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx]: Have gone back to individual fetches with a loading block on the reorder until this has been completed. This resolves any issues with using the wrong variant but also ensures users cannot continue without knowing the relevant information.
There is also the hope that production will be snappier than my local environment.
Having said all that, this is a rather niche use-case.

**Comment by benpratt77** [comment]
cursor review

File: apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx [hunk 2]
```diff
@@ -522,31 +545,70 @@ export default function OrderDialog({
...
+                (selection) => selection.productId,
+              ),
+            ),
+          ),
+        ];
+
+        const nextPicklistProductsById: Record<number, ProductSearch> = {};
+        if (picklistProductIds.length > 0) {
+          try {
+            const { productsSearch = [] } = await searchProducts({
```
Comment (benpratt77) [inline apps/storefront/src/pages/OrderDetail/components/OrderDialog.tsx]: Note: we fetch the picklist child's inventory with a direct searchProducts call in the load effect rather than the shared usePicklistInventory hook. This is to prevent the stuttering that occurred when the two fetches resolved separately and painted in two passes.

**Review by declankirk — APPROVED** [review]

**Review by bc-jeffteng — APPROVED** [review]
LGTM

**Review by bc-veer — APPROVED** [review]

File: apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx [hunk 3]
```diff
@@ -312,6 +345,29 @@ export default function OrderCheckboxProduct(props: OrderCheckboxProductProps) {
                     />
                   </Box>
                 )}
+                {picklistBackorderRows.map(({ selection, backorderFields: selectionFields }) => (
```
Comment (bc-lulu-yang) [inline apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx:348]: [Nit]

Wondering if here we can reuse the shared component `PicklistBackorderMessages`. The structure looks similar

Comment (benpratt77) [reply apps/storefront/src/pages/OrderDetail/components/OrderCheckboxProduct.tsx:348]: There is a slight difference in the UI here, reusing the PicklistBackorderMessages would left align the qty and a few other fields.
Whilst we could make this work, it would mean piping through a few sx changes for a single usage.
For now I think this is fine as is. If we have another use of this though I could be more convinced.

**Comment by bc-launchbay** [comment]
*b2b-buyer-portal* deployed to Staging GCS Upload
success [20260803035244](https://launchbay.bigcommerce.net/projects/515/releases/293479) by Launchbay

**Comment by bc-launchbay** [comment]
*b2b-buyer-portal* deployed to Production GCS Upload
success [20260803035244](https://launchbay.bigcommerce.net/projects/515/releases/293479) by Launchbay
